src/Fantomas/TokenParser.fs

```diff
@@ -90,91 +89,213 @@ let private createHashToken lineNumber content fullMatchedLength offset =
+
+        let (|TripleQuoteChars|_|) v =
+            match v with
+            | DoubleQuoteChar, DoubleQuoteChar, DoubleQuoteChar -> Some()
+            | _ -> None
+
+        let (|OpenParenChar|_|) = equalsChar '('
+        let (|AsteriskChar|_|) = equalsChar '*'
+        let (|NoCloseParenChar|_|) = differsFromChar ')'
+        let (|NewlineChar|_|) = equalsChar '\n'
+        let (|HashChar|_|) = equalsChar '#'
+        let (|BackSlashChar|_|) = equalsChar '\\'
+        let (|NoBackSlashChar|_|) = differsFromChar '\\'
+        let (|CloseParenChar|_|) = equalsChar ')'
+        let isSpace = (=) ' '
+
+        let processLine (hashContent: string) (lineContent: string) (lineNumber: int) (offset: int): Token list =
+            let hashContentLength = String.length hashContent
+
+            let tokens =
+                let defineExpressionWithHash = lineContent.Substring(hashContentLength)
+
+                if String.isNotNullOrEmpty defineExpressionWithHash
+                then tokenize [] [] defineExpressionWithHash
+                else []
+
+            tokens
+            |> List.map (fun t ->
+                let info =
+                    { t.TokenInfo with
+                          LeftColumn = t.TokenInfo.LeftColumn + hashContentLength
+                          RightColumn = t.TokenInfo.RightColumn + hashContentLength }
+
+                { t with
+                      LineNumber = lineNumber
+                      TokenInfo = info })
+            |> fun rest ->
+                (createHashToken lineNumber hashContent offset)
+                :: rest
+
+        let sourceLength = String.length sourceCode
+        // stop scanning the source code three characters before the end
+        // three because of how triple quote string are opened and closed
+        // the scan looks three characters ahead (zero, plusOne, plusTwo)
+        // and sometimes also two characters behind (minusTwo, minusOne)
+        // In theory there is will also never be any new hash detect inside the last three characters
+        let lastIndex = sourceLength - 3
+        // check if the current # char is part of an define expression
+        // if so add to defines
+        let captureHashDefine (state: SourceCodeParserState) idx =
+            let lastNewlineIdx =
+                Seq.tryHead state.NewlineIndexes
+                |> Option.defaultValue -1
+
+            let leadingCharactersBeforeLastNewlineAreSpaces =
+                let take = Math.Max(idx - lastNewlineIdx - 1, 0)
+
+                sourceCode
+                |> Seq.skip (lastNewlineIdx + 1)
+                |> Seq.take take
+                |> Seq.forall isSpace
+
+            if leadingCharactersBeforeLastNewlineAreSpaces then
+                let skip =
+                    if lastNewlineIdx = -1 then 0 else lastNewlineIdx + 1 // zero when the source starts with an #
+
+                let currentLine =
+                    sourceCode
+                    |> Seq.skip skip
```

Comment:
The use of `Seq.skip` always terrifies me, because it makes it very easy to write things which are quadratic in complexity; but strings are easy to seek around in, so that's fine.